pyangbind/lib/serialise.py

```diff
@@ -444,6 +444,112 @@ def generate_ietf_tree(obj, parent_namespace=None, flt=False, with_defaults=None
+            base_mod_cls = getattr(parent, safe_name(yang_base))
+            tmp = base_mod_cls(path_helper=False)
+
+            if path_helper is not None:
+                # check that this path doesn't already exist in the
+                # tree, otherwise we create a duplicate.
+                existing_objs = path_helper.get(tmp._path())
+                if len(existing_objs) == 0:
+                    obj = base_mod_cls(path_helper=path_helper, extmethods=extmethods)
+                elif len(existing_objs) == 1:
+                    obj = existing_objs[0]
+                else:
+                    raise pybindJSONUpdateError("update was attempted to a node that " + "was not unique")
+            else:
+                # in this case, we cannot check for an existing object
+                obj = base_mod_cls(path_helper=path_helper, extmethods=extmethods)
+
+        for child in d.getchildren():
+            # separate element namespace and tag
+            qn = etree.QName(child)
+            namespace, ykey = qn.namespace, qn.localname
+
+            # need to look up the key in the object to find out what type it should be,
+            # because we can't tell from the XML structure
+            attr_get = getattr(obj, "_get_%s" % safe_name(ykey), None)
+            if attr_get is None:
+                raise AttributeError("Invalid attribute specified (%s)" % ykey)
+            chobj = attr_get()
+
+            if chobj._yang_type == "container":
+
+                if hasattr(chobj, "_presence"):
+                    if chobj._presence:
+                        chobj._set_present()
+
+                pybindIETFXMLDecoder.load_xml(
+                    child, None, None, obj=chobj, path_helper=path_helper, extmethods=extmethods
+                )
+
+            elif chobj._yang_type == "list":
+                # TODO: development assertions, not sure how to resolve these yet
+                assert chobj._keyval is not False, "keyless list?"
+
+                # we just need to find the key value to add it to the list
+                key_parts = []
+                add_kwargs = {}
+                for pkv, ykv in zip(chobj._keyval.split(" "), chobj._yang_keys.split(" ")):
+                    add_kwargs[pkv] = child[ykv]
+                    key_parts.append(str(child[ykv]))
+                key_str = " ".join(map(str, key_parts))
+                if key_str not in chobj:
+                    nobj = chobj.add(**add_kwargs)
+                else:
+                    nobj = chobj[key_str]
+
+                # now we have created the nested object element, we add other members
+                pybindIETFXMLDecoder.load_xml(
+                    child, None, None, obj=nobj, path_helper=path_helper, extmethods=extmethods
+                )
+
+            elif hasattr(chobj, "_pybind_generated_by") and chobj._pybind_generated_by == "TypedListType":
+                chobj.append(str(child))
```

Comment:
Doesn't this case every type to string within a `TypedListType`? Does it work with `e.g., leaf-list foo { type union { type uint32; type string; }}`? I'd expect that in this case we can have an XML payload that is simply a `uint32`, whereas this looks like everything would be added to the list as a string.

Reply:
Great point. This turns out to be rather tricky, particularly with XML, as there is no real typing information.. this isn't a problem with the YANG as a schema.. except in the case of `union`s.

I've had an initial stab by relying on the lxml guess on matching Python type, which I think is a _reasonable_ assumption for minimal effort. I've added a fallback to stringify if and only if string is an accepted type. (The main use-case for this is when you have modelled a leaf as a string but are actually storing ints or floats (e.g. vlan ID, or some QoS priority perhaps..?)

I would like to suggest we go with this and have a separate task to improve and robustify following proper spec (e.g. section 9 RFC 7950).